### DISCUSSION TITLE
improve electron support

### DISCUSSION THREAD

**Discussion by BlackHole1** [pnpm/pnpm#5146] [General]
Currently, if an electron application uses pnpm, it must use `node-linker=hoisted`.See: https://github.com/electron-userland/electron-builder/issues/6289#issuecomment-1042620422

However, it would be better to have `electron-builder` / `electron-packager` / `electron-forge` support pnpm directly.

This is not easy to do, however, as pnpm's unique node_modules directory structure makes finding dependencies and child dependencies somewhat difficult.

Also, as I am a member of both electron and pnpm, I would like to make some efforts in this direction.

If this issue is resolved, it will benefit both the electron and pnpm communities and help the development of the community.

But I may not be able to do it well on my own and I need some help.

**Comment by BlackHole1** [comment]
@pnpm/collaborators I would like to know what you all think about this.

To be honest I don't know how to write a pnpm dependency and child dependency lookup tool, if someone could help me write a similar tool it would be very helpful, I would optimize it and rewrite it in Golang (the electron-builder dependency lookup is written in Golang)

**Reply by aparajita** [reply]
Wouldn't it be sufficient to list the top level packages, then for each one do `pnpm why -r`?

Remember there is are `--json` and `--parseable` flags which will help.

**Reply by BlackHole1** [reply]
Thank you for your reply, I made an attempt, however, I was unable to enumerate the child's dependencies.

**Reply by zkochan** [reply]
I don't understand what the issue is. How is the lookup implemented for node-linker=hoisted? It should work the same for the symlinked node_modules as well. The only difference would be that symlinks should be resolved to their real locations before resolving dependencies.

**Reply by BlackHole1** [reply]
I found a similar discussion: https://github.com/nodejs/node/discussions/37509

It seems that there may be more than just the problem of finding dependencies, but also the problem of symlinks.

**Comment by BlackHole1** [comment]
As far as I know, some electron packagers do not handle dependency lookups properly.
See: https://github.com/develar/app-builder/blob/master/pkg/node-modules/nodeModuleCollector.go#L182-L216 (electron-builder)

**Reply by BlackHole1** [reply]
So in my initial thinking, it was only necessary to get it to resolve the pnpm dependency structure correctly to make it work properly.
This is also the "how to find dependencies and child dependencies" problem I mentioned above.
But based on https://github.com/nodejs/node/discussions/37509, I found that even if I could handle the pnpm structure correctly, it didn't work. Because of the symlinks problem.

**Comment by BlackHole1** [comment]
I have come up with an idea, please correct me if there are problems.

1. Implement a script that works basically like `node-linker=hoisted`, but it doesn't do any disk writes. It would return a directory structure similar to the one we have after `node-linker=hoisted` as a result.
2. The electron package tool will depend on or reimplement this script and use its return value as the structure of `node_modules` when electron packages.

Pseudocode:
```ts
const hoisted = () => {
    // return `node-linker=hoisted` result;
}

const resovleNodeModules = () => {
    for (item of hoisted()) {
        cp(item.realPath, `electron/dist/node_modules/${item.hoistedPath}`)
    }
}

type Item {
   realPath: string; // .pnpm/got@12.3.0/node_modules/got/
   hoistedPath: string; // got
}
```

**Reply by zkochan** [reply]
I don't understand why you need a fake hoisted node_modules and how it is better than just simply using node-linker=hoisted.

**Reply by BlackHole1** [reply]
Because there are many people who want to keep the existing structure of pnpm and not use `node-linker=hoisted`

See: https://github.com/electron-userland/electron-builder/issues/6289#issuecomment-1049655548

**Comment by beyondkmp** [comment]
@BlackHole1 
I have written a library to support this functionality. It can read production dependencies from a non-hoisted pnpm setup and convert them into a hoisted mode.

https://github.com/beyondkmp/NodeModuleCollector
